ci_scripts/utils/configuration.py

```diff
@@ -0,0 +1,275 @@
+    MASTER_BRANCH = 'master'
+    RELEASE_BRANCH_PATTERN = r"^release.*$"
+    REMOTE_ALIAS = 'origin'
+    LOGGER_FORMAT = '%(levelname)s: %(message)s'
+    BOT_USERNAME = 'Monty Bot'
+    BOT_EMAIL = 'monty-bot@arm.com'
+    ORGANISATION = 'Arm Mbed'
+    ORGANISATION_EMAIL = 'support@mbed.com'
+
+    def _fetch_value(self, key: str) -> Any:
+        try:
+            return getattr(self, key)
+        except AttributeError:
+            self._raise_undefined(key)
+
+
+class EnvironmentConfig(GenericConfig):
+    """Configuration set in environment variables.
+
+    This also uses dotEnv mechanism.
+    """
+
+    def __init__(self):
+        """Constructor."""
+        dotenv.load_dotenv(
+            dotenv.find_dotenv(usecwd=True, raise_error_if_not_found=False))
+
+    def _fetch_value(self, key: str) -> Any:
+        environment_value = os.getenv(key)
+        if not environment_value:
+            self._raise_undefined(key)
+        return environment_value
+
+
+class FileConfig(GenericConfig):
+    """Configuration set in toml file."""
+
+    CONFIG_SECTION = 'ProjectConfig'
+    PATH_TOKEN = {'DIR', 'ROOT', 'PATH'}
+    CONFIG_FILE_NAME = 'pyproject.toml'
+
+    def __init__(self, file_path: str = None) -> 'FileConfig':
+        """Constructor.
+
+        Args:
+            file_path: path to the toml configuration file.
+        """
+        self._file_path: str = file_path
+        self._config: dict = dict()
+        self._has_loaded: bool = False
+        self._found_file: bool = False
+
+    def _adjust_path_values(self, variable_name: str, value: str) -> str:
+        """Works out the correct values for path variables.
+
+        Paths in the configuration file are relative to the configuration
+        file location. This method ensures the path values are therefore
+        evaluated properly
+        Args:
+            variable_name: name of the variable in the configuration file
+            value: variable value
+
+        Returns:
+            a valid path or the value unchanged if the variable is not a path
+
+        """
+        if not self._found_file:
+            return value
+        for token in FileConfig.PATH_TOKEN:
+            if token in variable_name:
+                value = os.path.realpath(
+                    os.path.join(
+                        os.path.dirname(self._file_path),
+                        value
+                    )
+                )
+                break
+        return value
+
+    def _look_for_config_file_walking_up_tree(self):
+        try:
+            self._file_path = find_file_in_tree(
+                FileConfig.CONFIG_FILE_NAME, top=True)
+            self._found_file = True
+        except FileNotFoundError as e:
+            logger.warning(e)
+            self._found_file = False
+
+    def _find_config_file(self):
+        if self._found_file:
+            return
+        if self._file_path and os.path.exists(self._file_path):
+            self._found_file = True
```

Comment:
I don't understand this bit. Why `self._found_file` is set to `True` and below in the `try` block it's set to `True` again? What's the case here?

This class has a bit too much internal state. I don't see how `self._found_file` and `self._loaded` would ever be needed on their own. The less your methods know about internal state and can operate on their own, the better. :)

How about something along the lines of:
```
class FileConfig(GenericConfig):
    def __init__(self, file_path: str = None) -> 'FileConfig':
        self._file_path: str = file_path
        self._config: Optional[dict] = None

    def _load_config_from_file(self):
        # orchestrate tree walking and trying to find the toml
        # ...
        return toml.load(f).get(FileConfig.CONFIG_SECTION, dict())

    def _get_config(self):
        # note, that this is the only place that cares if config was loaded or not,
        # no need for `_found_file` or `_has_loaded`
        if not self._config:
            self._config = self._load_config_from_file()
        return self._config

    def _fetch_value(self, key: str):
        config = self._get_config()
        try:
            value = config[key]
        except KeyError:
            self._raise_not_found(key)

        return self._make_path_absolute_to_toml(key, value)
```

Reply:
made the changes